uefi: serial: various improvements to doc, read(), and write()

phip1611 opened the pull request:
This improves multiple things for the serial IO protocol. Please review commit by commit.

I tested this in a VM as well as on real hardware. The changes make the Serial IO protocol more usable and also help to better cope with non-spec compliant implementations.


## Checklist
- [ ] Sensible git history (for example, squash "typo" or "fix" commits). See the [Rewriting History](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) guide for help.
- [ ] Update the changelog (if necessary)

nicholasbishop commented on uefi/CHANGELOG.md:
```diff
@@ -60,7 +60,8 @@
   - Removed `Directory::read_entry_boxed_in`
-
+- Fixed `Serial::read` to properly read only as many bytes as it can report
+  that `n` to the caller.
```
This ended up in the wrong version (0.36 instead of unreleased)

phip1611 committed:
uefi-raw: serial: fix mutability

We are logically modifying state, so this should be mut.

nicholasbishop commented on uefi/src/proto/console/serial.rs:
```diff
@@ -81,15 +84,43 @@ impl Serial {
+            // UEFI was either not able to fill the whole buffer in the specified
+            // timeout, but nevertheless read some data, or there was too much
+            // data.
+            Status::TIMEOUT => Ok(buffer_size),
```
I'm not sure we should swallow the `Timeout` error like this, end users might have good reason to handle timeout differently.

It's a bit ugly, but what if we instead added an output arg like `num_bytes_read: &mut usize`, and document that it will be set correctly in both the success and error cases?

phip1611 replied:
> end users might have good reason to handle timeout differently.

Fair concern, there might be use-cases for that. Let me think about it again.

(I was working on a fun project over the holidays and I adapted it to mainly work for my use-case)

nicholasbishop commented on uefi/src/proto/console/serial.rs:
```diff
@@ -99,10 +130,8 @@ impl Serial {
     pub fn write(&mut self, data: &[u8]) -> Result<(), usize> {
         let mut buffer_size = data.len();
-        unsafe { (self.0.write)(&mut self.0, &mut buffer_size, data.as_ptr()) }.to_result_with(
-            || debug_assert_eq!(buffer_size, data.len()),
```
That would seem to conflict with this function's docstring: "This operation will block until the data has been fully written or an error occurs." If the docstring is correct, than the assert can stay, otherwise we should fix the docstring as well. I find the UEFI spec a little ambiguous in its language here. "EFI_SUCCESS -- The data was written" could be taken to mean _all_ the data was written, but I'm not sure. Did you observe that calling `write` multiple times is needed in some real-world test?

phip1611 committed:
uefi: serial: improve clarity over timeout in SerialIoMode

phip1611 committed:
uefi-raw: extend support for revision in SERIAL_IO_PROTOCOL

phip1611 committed:
uefi: serial: add support for revision

phip1611 committed:
uefi: serial: improve overall documentation

phip1611 committed:
uefi: serial: improve read() and write()

Next to documentation improvements, I replaced the debug assertions with
runtime assertions. If the underlying protocol implementation doesn't
behave as expected, we should at least fail hard.

I tested this with OVMF and also on real hardware.

phip1611 committed:
uefi: serial: fix core::fmt::Write impl

This is the more expected and natural behavior.

phip1611 committed:
uefi: serial: add read_exact() and write_exact()

phip1611 committed:
uefi: serial: read_to_vec()

nicholasbishop reviewed (commented):
Adding a few more comments. I haven't fully checked the rest though; I think it would be good to try to split this PR up since there are a lot of changes to consider here. Perhaps one PR for the uefi-raw changes, one for the docstring improvements to existing functions, one each for the changes to read and write, and a couple more for read_to_vec/write_exact.
